CSVDiff to SchemaDiff

milljm opened the pull request:
Use DeepDiff for diffing routines in the TestHarness.

Unable to push to original repository, creating a new PR.

Refs# 23522

permcody committed:
Convert CSVDiff to SchemaDiff

closes #23522

Co-authored-by: Cody Permann <codypermann@gmail.com>

permcody committed:
Fixing/Improving CSVDiffer operation

Commit:
Bring back some previous features

Allow working with Custom Comparison files
Allow overrides in test spec files
Fix: Make unittest abide by default MOOSE_TERM_FORMAT types
Display more useful errors when something diffs

Refs #23522

Commit:
Create Exception test for SchemaDiff

Stip leading spaces in CSV field names, e.g:
(test/tests/outputs/csv/gold/csv_align_out.csv).

Create a mock exception test for SchemaDiff.

Modify default MOOSE_TERM_COLS, as I found if set low enough, some
unittests fail when regexing for lengthy caveats.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1577730/) on d01d118 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24633/site)

This comment will be updated on new commits.

Commit:
Move custom operator out of testers

Create a DeepDiff Comparison Class outside of TestHarness testers,
to allow tests to be skipped if deepdiff is not available.

Type change where necessary.

Add --re as scenario when not to overwrite TestHarness previous
results.

Commit:
Remove unnecessary import

found a test that requires matplotlib.

lindsayad commented on test/tests/test_harness/csvdiff_tests:
```diff
@@ -95,4 +95,11 @@
+    input = csvdiff_comparison_custom_column.i
+    csvdiff = csvdiff_comparison_custom_column_out.csv
+    custom_columns = time
+    requirement = 'CSVDiff shall compare only a specific column'
```
```suggestion
    requirement = 'The system shall be able to diff CSV files for only a specific column.'
```

lindsayad reviewed (changes requested):
Maybe a person who does python more than I do should review this 😆

Commit:
Getting a better handle on DeepDiff

Returning bools on give_up_diffing does not work as I
had originally thought...

You are instead telling DeepDiff if it can continue to
diff on the object or not (hint: read the method name).
Where I was getting hung up was from DeepDiff not
reporting a diff no matter what bool I was returning. I
later found this is due to 'exclude_paths'.

Concerning test: reporters/constant_reporter

To summarize:

return True:   DeepDiff will no longer try to perform a
               diff: pass (if you do not also set a custom
               report)

return False:  DeepDiff allowed to continue, but will
               make use of 'exlude_paths' set forth by
               test specs: also resulting in a pass

DeepDiff Ref: https://zepworks.com/deepdiff/current/custom.html#define-a-custom-operator

Refs #23522

Commit:
Fix isinstance

I used wrong type to compare against.
LOL. Fix test_InstallType one more time.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1587226/) on 263b56d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4278fb</th>
      <th colspan=4>#24633 263b56</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.40%</td>
      <td>85.40%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88878</td>
      <td>88883</td>
      <td>+5</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15197</td>
      <td>15192</td>
      <td>-5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24633/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24633/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24633/coverage/xfem)

</details>



This comment will be updated on new commits.

Commit:
Refactor CSVDiff for loop, custom options

Pass individual rows to DeepDiff. As it turns out, you don't
want to send it overly populated dictionaries of lists. While
the end result is the same, a 2MB CSV can take ~26 minutes to
compare. Ref: https://github.com/pytest-dev/pytest/issues/8998

Refactor the nested for loop/custom option logic.

Fix Install Type unittest to work on Civet and Peoples machine
(mock objects ignore exported MOOSE_TERM_FORMAT variables).

Closes #23522

dschwen commented:
We don't have any tests that actually make sure that a tester that is meant to fail will fail, do we?

loganharbour commented:
> We don't have any tests that actually make sure that a tester that is meant to fail will fail, do we?

Nope. But we will now.
